### PR TITLE
Adds support to filter AZD templates by tags

### PR DESCRIPTION
**Pull request by wbreza**
## Feature Updates

- Adds `tags` to JSON structure for AZD template sources
- Adds (`--filter, -f`) option to `azd template list`
- Adds (`--filter`) options to `azd init`
- Displays tags in `azd template show`

## Filter templates during `azd init`

When specifying `--filter, -f` argument to `azd init` AZD will automatically enter template list mode showing only templates that match the specified tags.

image

## Filter templates during `azd template list`

When specifying `--filter, -f` argument to `azd template list` AZD will filter the template results to show only those templates that have matching tags.

image

## Display Tags in `azd template show <template>`

Tags will now show along with other template metadata.

image

### CODE REVIEW

**Comment by rajeshkamal5050** [comment]
Currently `azd template list` lists all 106. Any default grouping we could do? based on languages? or any other category?

@ellismg @kristenwomack @wbreza @Austinauth

**Comment by kristenwomack** [comment]
Default grouping by language gives better distribution today, with only a few outliers. Grouping by scenario or app topic could be cool, but we would need to make sure the tag exists on each template in order for it to work properly. Currently, filtering by "topics" doesn't yield the same number as the total templates we have in the gallery. 

This would also be helpful to have some grouping in the VS Code extension to visually organize or "chunk" the large list into groups. Not knowing what to do with such a long list (esp in VS Code) could stop folks from getting to the next step.

**Comment by wbreza** [comment]
> Currently `azd template list` lists all 106. Any default grouping we could do? based on languages? or any other category?
> 
> @ellismg @kristenwomack @wbreza @Austinauth

Some grouping and UX revision would be good... Lets take that as a separate feature.  This gives you better filtering which will give you a smaller list.

File: cli/azd/cmd/init.go [hunk 1]
```diff
@@ -72,6 +73,12 @@ func (i *initFlags) Bind(local *pflag.FlagSet, global *internal.GlobalCommandOpt
 		"b",
 		"",
 		"The template branch to initialize from. Must be used with a template argument (--template or -t).")
+	local.StringSliceVar(
+		&i.templateTags,
+		"tags",
```
Comment (vhvb1989) [inline cli/azd/cmd/init.go:78]: @kristenwomack @wbreza , did you maybe considered `filter`  or `filter-tags` ?

Without reading at the help, the flag by itself is a little confusing (to me, at least).  I think I am totally bias by the `tag` flag from Docker cli and the git tag command, but, reading `tags` makes me thing on an action-flag that will create/write something.   I can't help thinking `azd init --tags foo` will add a tag `foo` to my project

Comment (wbreza) [reply cli/azd/cmd/init.go:78]: I'm open to suggestions on argument naming.

Comment (kristenwomack) [reply cli/azd/cmd/init.go:78]: I like `filter` or `filter-tags` a lot. It immediately makes my brain think of reducing a list to something specific.

Comment (wbreza) [reply cli/azd/cmd/init.go]: Updated to `--filter, -f`.  Today we only support flags but in the future we could expand to filter on other options so keeping this more generic to `filter` feels like the way to go in my opinion.

File: cli/azd/cmd/init.go [hunk 2]
```diff
@@ -162,7 +170,7 @@ func (i *initAction) Run(ctx context.Context) (*actions.ActionResult, error) {
 	}
 
 	var initTypeSelect initType
-	if i.flags.templatePath != "" {
+	if i.flags.templatePath != "" || len(i.flags.templateTags) > 0 {
```
Comment (vhvb1989) [inline cli/azd/cmd/init.go:173]: Should we add validation about using `--from-code` and `--filter` as an error?

**Review by vhvb1989 — APPROVED** [review]

**Review by ellismg — APPROVED** [review]
LGTM

**Comment by irinasmoke** [comment]
> > Currently `azd template list` lists all 106. Any default grouping we could do? based on languages? or any other category?
> > @ellismg @kristenwomack @wbreza @Austinauth
> 
> Some grouping and UX revision would be good... Lets take that as a separate feature. This gives you better filtering which will give you a smaller list.

Is there a new issue created for this? I was checking out the filter feature and couldn't find it at first because I didn't know about the --filter flag. I was expecting there to be some sort of dynamic filtering, like we have for the long list pattern, so that I could see all the templates but then go into search mode. I worry that the filter flag is not very discoverable, especially for new users, so I'm wondering if we could enable something that the user can be aware of filtering after they run the base command?

**Comment by Austinauth** [comment]
I do think that no matter what we do, we have to get a handle on the length of the names to make the template select experience more parsable in the CLI. I also feel like at this point in time, adding the url to the template select has only made the list harder to scan. I would not be opposed to going back to the simple list we originally had (no URL) -- at least until we have a chance rework the pattern. 

The new `template show <template name>` will give users a way to see the URL. I know its not very discoverable, but at least it wont be broken.

**Pull request merged**